flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java

```diff
@@ -734,6 +744,18 @@ public void requestHeartbeat(ResourceID resourceID, Void payload) {
 		}
 	}
 
+	private void registerSlotAndTaskExecutorMetrics() {
+		jobManagerMetricGroup.gauge(
+			TASK_SLOTS_AVAILABLE_METRIC_NAME,
+			() -> (long) slotManager.getNumberFreeSlots());
```

Comment:
@zentol There may be problems with read inconsistencies here. But I think the impact is small:

* If RM and Dispatcher are single-threaded, there will be no concurrent modification problems here;
* The indicator is periodically extracted (inconsistent is only a short period of one cycle) and its value does not play any decision role in Flink (should be tolerable?)
* Considering the frequency of metric extraction, if we provide related synchronization mechanisms, performance will be reduced and complexity will be increased.

What do you think?

Reply:
I think it should be okish to call these methods from a different thread. All of them simply return the size of a `HashMap`. In the worst case, we would report a little bit stale data. The main danger I see is that the implementations of `SlotManager` might change, which makes this no longer hold true.

If we wanted to make it "thread-safe", then we would need to make the access call inside of the `MainThreadExecutor`:
```
jobManagerMetricGroup.gauge(
	MetricNames.TASK_SLOTS_AVAILABLE,
	() -> callAsync(() -> slotManager.getNumberFreeSlots(), timeout).get());
```

However, this might be an overkill at the moment.

Reply:
yes, my concern was primarily over the long-term. This also becomes relevant if new metrics are added, which will follow a similar pattern as existing ones that don't worry about concurrency.

However, @tillrohrmann's solution isn't even overkill enough, as it could result in exceptions being thrown by metrics (in case of a timeout) which generally shouldn't happen. So we'd have to cache the previous result and return that in case of a timeout.

Reply:
@zentol So, I should accept this solution?

> So we'd have to cache the previous result and return that in case of a timeout.

Reply:
I think this problem might occur at other places as well. Thus, at some point it might be helpful to either add some tooling to do exactly what @zentol proposed or to make the metric system work with asynchronous metrics.

Reply:
@tillrohrmann so, Your suggestion is that we will not solve it in this PR, right?

Reply:
Solving this issue would be overkill for this PR, I agree.